language/diem-framework/modules/Vote.move

```diff
@@ -0,0 +1,377 @@
+    /// per address.
+    const MAX_BALLOTS_PER_PROPOSAL_TYPE_PER_ADDRESS: u64 = 256;
+
+    /// The provided timestamp(s) were invalid
+    const EINVALID_TIMESTAMP: u64 = 1;
+    /// The address already contains has the maximum of ballots allowed
+    /// MAX_BALLOTS_PER_PROPOSAL_TYPE_PER_ADDRESS
+    const ETOO_MANY_BALLOTS: u64 = 2;
+    /// Ballot with the provided id was not found
+    const EBALLOT_NOT_FOUND: u64 = 3;
+    /// Proposal details in the vote do not match the proposal details
+    /// in the ballot
+    const EBALLOT_PROPOSAL_MISMATCH: u64 = 4;
+    /// Voter not allowed to vote in the ballot
+    const EINVALID_VOTER: u64 = 5;
+    /// current timestamp > ballot expiration time
+    const EBALLOT_EXPIRED: u64 = 7;
+    /// Voter has already voted in this ballot
+    const EALREADY_VOTED: u64 = 8;
+
+    /// A constructor for BallotID
+    public fun new_ballot_id(
+        counter: u64,
+        proposer: address,
+    ): BallotID {
+        BallotID {
+            counter,
+            proposer,
+        }
+    }
+
+    /// A constructor for WeightedVoter
+    public fun new_weighted_voter(
+        weight: u64,
+        voter: vector<u8>,
+    ): WeightedVoter {
+        WeightedVoter {
+            weight,
+            voter,
+        }
+    }
+
+    /// Create a ballot under the signer's address and return the `BallotID`
+    public fun create_ballot<Proposal: store + copy + drop>(
+        ballot_account: &signer,
+        proposal: Proposal,
+        proposal_type: vector<u8>,
+        num_votes_required: u64,
+        allowed_voters: vector<WeightedVoter>,
+        expiration_timestamp_secs: u64
+    ): BallotID acquires Ballots, BallotCounter {
+        let ballot_address = Signer::address_of(ballot_account);
+
+        assert(DiemTimestamp::now_seconds() < expiration_timestamp_secs, Errors::invalid_argument(EINVALID_TIMESTAMP));
+
+        if (!exists<BallotCounter>(ballot_address)) {
+            move_to(ballot_account, BallotCounter {
+                counter: 0,
+            });
+        };
+        if (!exists<Ballots<Proposal>>(ballot_address)) {
+            move_to(ballot_account, Ballots<Proposal> {
+                ballots: Vector::empty(),
+                create_ballot_handle: Event::new_event_handle<CreateBallotEvent<Proposal>>(ballot_account),
+                remove_ballot_handle: Event::new_event_handle<RemoveBallotEvent>(ballot_account),
+                voted_handle: Event::new_event_handle<VotedEvent>(ballot_account),
+                ballot_approved_handle: Event::new_event_handle<BallotApprovedEvent>(ballot_account),
+            });
+        };
+
+        let ballot_data = borrow_global_mut<Ballots<Proposal>>(ballot_address);
+
+        // Remove any expired ballots
+        gc_internal<Proposal>(ballot_data);
+        let ballots = &mut ballot_data.ballots;
+
+        assert(Vector::length(ballots) < MAX_BALLOTS_PER_PROPOSAL_TYPE_PER_ADDRESS, Errors::limit_exceeded(ETOO_MANY_BALLOTS));
```

Comment:
What do you think about holding a count of total number of ballots outstanding under the account's `BallotCounter` resource and then bounding off of that instead of bounding the number of each proposal type?

Reply:
I think this is better because it's easy to run gc against a specific proposal type when creating a new ballot.

If `Proposal1` type has filled up all the slots with expired Ballots and the proposer tries to create a new ballot with `Proposal2` type, it's not possible for the create_ballot function to gc the `Proposal1` ballots.

Here we can automatically gc ballots of the same type and ensure that a ballot can always be created of a specific type as long as there aren't too many of of that type lying unexpired.